### PR TITLE
wrap `admin` and `preview` in `pan-domain-authentication` 🐼 

### PR DESCRIPTION
**Pull request by twrichards**
### Prerequisites
- [x] Needs https://github.com/guardian/pan-domain-authentication/pull/137 to be merged and released.
- [x] So that this PR doesn't break viewer (the tool via which vast majority of editorial users acutally consume preview) https://github.com/guardian/editorial-viewer/pull/136 needs to be merged & deployed
- [x] infra changes to allow access to the `pan-domain-auth-settings` bucket by `preview` and `admin` servers are in https://github.com/guardian/platform/pull/1589

## What is the value of this and can you measure success?
Prior to this PR the `admin` and `preview` apps within frontend were wrapped in plain Google auth (since they require a Guardian login to use) - this leaves us at the mercy of Google outages (as [occurred recently on 5th March](https://www.theguardian.com/technology/2024/mar/05/facebook-instagram-outages-disruption-meta-google) for example). This PR replaces Google auth with our own [pan-domain-authentication](https://github.com/guardian/pan-domain-authentication) which wraps Google auth with a [handy extra feature](https://github.com/guardian/login.gutools#emergency-access-when-google-auth-is-down) to make us resilient to such outages. 

This also tees us up nicely for adding permissions to `admin` and `preview` in https://github.com/guardian/frontend/pull/27078

## What does this change?
Overall this is NET reduction in code 🎉  and a reasonable amount of the diff is splitting/moving things around. A summary of the changes is as follows...

- add `nginx/nginx-mapping.yml` for use with [`dev-nginx`](https://github.com/guardian/dev-nginx) so that `admin` and `preview` can run on `frontend.local.dev-gutools.co.uk` and `preview.local.dev-gutools.co.uk` respectively - so that the pan-domain-authentication actually works (won't work when running purely on localhost)
- renames `GoogleAuthFilters` in common, to the more fitting `GuardianAuthWithExemptions`
- replaces the `com.gu.play-googleauth` lib with `pan-domain-auth-play_3-0` (and removes the now unnecessary `com.gu.play-secret-rotation` and `com.gu.play-secret-rotation`)
- initialises instances of `GuardianAuthWithExemptions` in `admin` and `preview`'s `AppLoader`s 
- within `GuardianAuthWithExemptions`
   -  initialises the `PanDomainAuthSettingsRefresher` which is central to how panda works
   - provides a `filter` property which implements `play.api.mvc.Filter` (for use in admin and preview's lists of filters similar to how `GoogleAuthFilters` was used before
   - leaves placeholder for adding permission checking - see https://github.com/guardian/frontend/pull/27078 and https://github.com/guardian/frontend/pull/27092
- removes `AdminAuthAction` replaces uses of it with regular `Action` since the filter should be enforcing auth more broadly
- strips out login buttons and endpoints from admin and preview since panda signs in / redirects automatically

## Screenshots
N/A

## Checklist

- [x] Tested locally, and on CODE if necessary
- [ ] Will not break dotcom-rendering
- [ ] Will not break our database – if updating CAPI, [updated and committed the database files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)
- [ ] Meets our accessibility [standards](https://github.com/guardian/recommendations/blob/e647ef695199ea3116ea20d827ef0f1364270a39/accessibility.md)
  - [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
  - [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#Keyboard)
  - [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#colour)

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -58,7 +58,7 @@ object Dependencies {
   val macwire = "com.softwaremill.macwire" %% "macros" % "2.5.7" % "provided"
   val mockito = "org.mockito" % "mockito-all" % "1.10.19" % Test
   val paClient = "com.gu" %% "pa-client" % "7.0.7"
-  val playGoogleAuth = "com.gu.play-googleauth" %% "play-v30" % "4.0.0"
+  val panDomainAuth = "com.gu" %% "pan-domain-auth-play_3-0" % "3.1.0"
   val playSecretRotation = "com.gu.play-secret-rotation" %% "play-v30" % "7.1.0"
   val playSecretRotationAwsSdk = "com.gu.play-secret-rotation" %% "aws-parameterstore-sdk-v1" % "7.1.0"
```
Comment (rtyley) [inline project/Dependencies.scala:63]: Note that deleting `common/app/conf/GoogleAuth.scala` in this PR means there's no remaining use of `com.gu.play-secret-rotation` in the project, so these two dependencies can be removed too!

I'm not sure what the rotation policy is on the panda keys, I do wonder how often they are rotated?!

Comment (twrichards) [reply project/Dependencies.scala]: great spot, thanks @rtyley 

I'll look into the panda key rotation.

File: admin/app/controllers/HealthCheck.scala [hunk 2]
```diff
@@ -1,13 +1,12 @@
...
+class HealthCheck(val controllerComponents: ControllerComponents) extends BaseController {
 
-class HealthCheck(wsClient: WSClient, val controllerComponents: ControllerComponents)(implicit
-    executionContext: ExecutionContext,
-) extends AllGoodCachedHealthCheck(
-      NeverExpiresSingleHealthCheck("/login"),
-    )(wsClient, executionContext)
+  def healthCheck(): Action[AnyContent] =
+    Action {
+      Ok("OK")
```
Comment (mxdvl) [inline admin/app/controllers/HealthCheck.scala:9]: 0K

File: admin/app/controllers/cache/ImageDecacheController.scala [hunk 3]
```diff
@@ -1,28 +1,22 @@
...
 import common.{GuLogging, ImplicitControllerExecutionContext}
-import controllers.admin.AdminAuthController
 import model.{ApplicationContext, NoCache}
-import play.api.http.HttpConfiguration
 import play.api.libs.ws.{WSClient, WSResponse}
-import play.api.mvc.Security.AuthenticatedRequest
 import play.api.mvc._
 
+import java.net.URI
+import java.util.UUID
```
Comment (mxdvl) [inline admin/app/controllers/cache/ImageDecacheController.scala:9]: Why change the order of these imports?

File: admin/app/controllers/cache/PageDecacheController.scala [hunk 4]
```diff
@@ -58,13 +50,15 @@ class PageDecacheController(
...
+  private def getSubmittedUrlPathMd5(request: Request[AnyContent]): Option[String] = {
     request.body.asFormUrlEncoded
       .getOrElse(Map.empty)
       .get("url")
       .flatMap(_.headOption)
       .map(_.trim)
-      .map(url => DigestUtils.md5Hex(new URI(url).getPath))
+      .map(new URI(_))
+      .map(_.getPath)
+      .map(DigestUtils.md5Hex)
```
Comment (mxdvl) [inline admin/app/controllers/cache/PageDecacheController.scala:61]: unroll please 🧵 👏

File: admin/app/views/admin_main.scala.html [hunk 5]
```diff
@@ -56,7 +56,7 @@
 
     </head>
     <body>
-        @admin_head(isAuthed)
+        @admin_head()
```
Comment (mxdvl) [inline admin/app/views/admin_main.scala.html:59]: does this still need to be a method being called? I don’t know much about Play 🤷

File: admin/test/TestAppLoader.scala [hunk 6]
```diff
@@ -1,6 +1,15 @@
 import app.FrontendComponents
 import play.api.ApplicationLoader.Context
 import play.api.BuiltInComponentsFromContext
+import play.api.mvc.EssentialFilter
+import org.scalatestplus.mockito.MockitoSugar
```
Comment (mxdvl) [inline admin/test/TestAppLoader.scala:5]: Mockito Sugar, yum 😋

File: admin/test/TestAppLoader.scala [hunk 7]
```diff
@@ -11,5 +20,12 @@ trait TestComponents {
 
 class TestAppLoader extends AppLoader {
   override def buildComponents(context: Context): FrontendComponents =
-    new BuiltInComponentsFromContext(context) with AppComponents with TestComponents
+    new BuiltInComponentsFromContext(context) with AppComponents with TestComponents with MockitoSugar {
+      private val mockAuth = mock[GuardianAuthWithExemptions]
+      when(mockAuth.filter).thenReturn(new Filter {
+        implicit val mat: Materializer = materializer
+        def apply(nextFilter: RequestHeader => Future[Result])(request: RequestHeader) = nextFilter(request) // no auth
```
Comment (mxdvl) [inline admin/test/TestAppLoader.scala:27]: What does `no auth` mean here? Are we passing an unauthenticated request to the next filter?

File: common/app/http/GuardianAuthWithExemptions.scala [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    override val mat: Materializer = outer.mat
+
+    private def doNotAuthenticate(request: RequestHeader) =
+      context.environment.mode == Mode.Test ||
+        (List(
+          new URL(authCallbackUrl).getPath, // oauth callback
+          "/assets",
+          "/favicon.ico",
+          "/_healthcheck",
```
Comment (mxdvl) [inline common/app/http/GuardianAuthWithExemptions.scala:77]: 🧑‍⚕️ 🩺

File: nginx/setup.sh [hunk 9]
```diff
@@ -11,8 +11,10 @@ SOURCE_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 
 dev-nginx setup-cert ${DOMAIN}
 dev-nginx link-config ${SOURCE_DIR}/frontend.conf
+dev-nginx setup-app ./nginx-mapping.yml # for preview and admin
 dev-nginx restart-nginx
 
 echo -e "💯 Done! You can now run frontend locally on https://${DOMAIN}"
+echo -e "You can also (separately) run admin and preview locally on https://frontend.local.dev-gutools.co.uk and https://preview.local.dev-gutools.co.uk respectively."
```
Comment (mxdvl) [inline nginx/setup.sh:18]: 🔐

File: nginx/frontend.conf [hunk 10]
```diff
@@ -1,44 +1,66 @@
...
+    proxy_set_header X-Real-IP $remote_addr;
+    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_set_header Host $http_host;
+    proxy_redirect default;
+    proxy_buffering off;
+  }
+
+  ssl_certificate     preview.local.dev-gutools.co.uk.crt;
+  ssl_certificate_key preview.local.dev-gutools.co.uk.key;
```
Comment (mxdvl) [inline nginx/frontend.conf:17]: How are these keys generated?

File: preview/app/http/PreviewContentSecurityPolicyFilter.scala [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package http
+
+import org.apache.pekko.stream.Materializer
+import play.api.mvc.{Filter, RequestHeader, Result}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+// This should be kept up to date with the prod configured in VCL (fastly-edge-cache repo).
+// We don't have fastly or a cache in front of `preview.gutools.co.uk` as it provisioned in CloudFormation in the `platform` repo (private):
+// https://github.com/guardian/platform/blob/main/provisioning/cloudformation/frontend.yaml#L824
```
Comment (mxdvl) [inline preview/app/http/PreviewContentSecurityPolicyFilter.scala:10]: If pointing to a specific line, should we make this a permalink?
```suggestion
// https://github.com/guardian/platform/blob/06c343b0c44ba5bdcd1d7b613aed5e896f62684a/provisioning/cloudformation/frontend.yaml#L824
```

File: preview/app/http/PreviewNoCacheFilter.scala [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package http
+
+import org.apache.pekko.stream.Materializer
+import play.api.mvc.{Filter, RequestHeader, Result}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+// OBVIOUSLY this is only for the preview server
+// NOT to be used elsewhere...
```
Comment (mxdvl) [inline preview/app/http/PreviewNoCacheFilter.scala:9]: Would `/** */`-style comments not offer better IDE support for where NOT to use this class?

File: preview/app/http/PreviewContentSecurityPolicyFilter.scala [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+class PreviewContentSecurityPolicyFilter(implicit val mat: Materializer, executionContext: ExecutionContext)
+    extends Filter {
+  override def apply(nextFilter: (RequestHeader) => Future[Result])(request: RequestHeader): Future[Result] =
+    nextFilter(request).map(
+      _.withHeaders(
+        "Content-Security-Policy" -> "default-src https:; script-src https: 'unsafe-inline' 'unsafe-eval' blob: 'unsafe-inline'; frame-src https: data:; style-src https: 'unsafe-inline'; img-src https: data: blob:; media-src https: data: blob:; font-src https: data:; connect-src https: wss: blob:; child-src https: blob:; object-src 'none'; base-uri https://*.gracenote.com",
```
Comment (mxdvl) [inline preview/app/http/PreviewContentSecurityPolicyFilter.scala:16]: Where does this come from?

Would it help to [construct this string from a human-readable list](https://github.com/guardian/fastly-edge-cache/pull/1028/files)?

**Review by mxdvl — APPROVED** [review]
Looks good 🐼 

Some questions and comments for my own education, but nothing blocking I could spot

File: admin/app/controllers/cache/ImageDecacheController.scala (hunk 3 above)
Comment (twrichards) [reply admin/app/controllers/cache/ImageDecacheController.scala:9]: I think that was intelliJ

File: admin/app/views/admin_main.scala.html (hunk 5 above)
Comment (twrichards) [reply admin/app/views/admin_main.scala.html:59]: probably not, but I didn't want to tinker too much

File: admin/test/TestAppLoader.scala (hunk 7 above)
Comment (twrichards) [reply admin/test/TestAppLoader.scala:27]: oh it just means in the test we're essentially saying the filter does nothing, so essentially 'yes' to your question

the tests that this change gets working, weren't testing auth - hence I think this is justified

File: nginx/frontend.conf (hunk 10 above)
Comment (twrichards) [reply nginx/frontend.conf:17]: by [`dev-nginx`](https://github.com/guardian/dev-nginx), see changes to `nginx/setup.sh`

File: preview/app/http/PreviewNoCacheFilter.scala (hunk 12 above)
Comment (twrichards) [reply preview/app/http/PreviewNoCacheFilter.scala:9]: moved this from elsewhere, I can make that change if you think this is important?

File: preview/app/http/PreviewContentSecurityPolicyFilter.scala (hunk 13 above)
Comment (twrichards) [reply preview/app/http/PreviewContentSecurityPolicyFilter.scala:16]: moved this from elsewhere, I can make that change now if you think this is important? 

I do actually plan on adding to this preview specific CSP pretty soon, as I plan/hope to rip-out the proxying of preview from viewer (see mention in https://github.com/guardian/editorial-viewer/pull/136) so will need to add a [`frame-ancestors`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors) directive, so can make it nicer then?

File: preview/app/http/PreviewNoCacheFilter.scala (hunk 12 above)
Comment (mxdvl) [reply preview/app/http/PreviewNoCacheFilter.scala:9]: If only moved not sure I have an opinion 🤷

**Pull request merged**